### PR TITLE
[qa_automation] Add "qa_script_output" method to safely get shell command output

### PR DESCRIPTION
**Pull request by jtzhao**
During our testing, sometimes curl becomes very slow, which makes "script_output" fail. Instead of fetching the script with curl, "qa_script_output" types the shell command into console using "type_string" api, saves the output into a temporary file, and write the output into serial console. At last, it uses "wait_serial" to get the output.

### CODE REVIEW

File: tests/qa_automation/qa_run.pm [hunk 1]
```diff
@@ -95,6 +95,30 @@ sub start_testrun {
     assert_script_run "/usr/share/qa/qaset/run/$testsuite-run.openqa";
 }
 
+# Safely run shell commands and get output
+sub qa_script_output {
+    my ($self, $cmd, $timeout) = @_;
+    my $random      = int(rand(999999));
+    my $output_file = "/tmp/SCRIPT_OUTPUT_$random";
+    $timeout = 90 unless defined $timeout;
```
Comment (okurz) [inline tests/qa_automation/qa_run.pm]:
```perl
$timeout //= 90;
```

**Comment by okurz** [comment]
Sorry, I am unsure about this one. I don't see how this code is special and if this shouldn't be handled on a higher level for every test.

**Comment by jtzhao** [comment]
This method is designed to solve several issues:
1.The **script_output** api uses curl to fetch the script and run it. But sometimes curl becomes quite slow, which results in timeout problem. To avoid this, **qa_script_output** uses **type_string** to input the shell command.
2.Serial console pollution. While you're waiting for the output to show up in serial console with **wait_serial**, if openQA is also printing something like error info into it, you'll get both the output and the error info. In other words, your script's output is polluted. To reduce such situations, qa_script_output adds SCRIPT_BEGIN_<random> to the beginning and SCRIPT_END_<random> to the end of the output.

Not sure if others need it. So it's only added in qa_run.pm.

**Comment by coolo** [comment]
using type_string instead of curl for short commands is something I had planned myself. It just makes sense to do in general - not just if curl is slow. But of course you have to find a good balance - but if the script is one line only -> use type_string.

Adding BEGIN and END of course makes the thing to type much longer - which would be a win for curl again, but I don't really see why you need that. Perhaps you can explain.

Anyway, I also think that the problems you face are generic enough to fix it at the right place, e.g. with options to script_output (like we have in type_string [to avoid the meta discussion again :]) specifying if to use curl or not (where the default could depend on the command type) and specifying if to wrap the output (because in most scenarios it doesn't matter) if there is more to the thing you check for.

**Comment by jtzhao** [comment]
Suppose we have a shell command which takes 10s to finish. Normally we'll type the command and use wait_serial to wait for the END string. During this period, if openQA prints some error message into serial console, wait_serial will also capture those irrelevant info. We met this before during our testing. That's why I added BEGIN string at the beginning.

**Comment by jtzhao** [comment]
@coolo Hi, this function is only for qa_run.pm. Actually we're already using this approach for a few weeks. This time I'm just encapsulating them into a method, so that I can invoke it elsewhere.

**Comment by coolo** [comment]
yes. And I'm asking you to take it one step further to make it available for other testers too.

**Comment by coolo** [comment]
please really consider it

**Comment by jtzhao** [comment]
@coolo Sure. I can do that. So we need **script_output** api to support lightweight tasks such as a one-line shell command by adding an option, right?

**Pull request merged**